util/tracer/tracer.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tobias Schottdorf (tobias.schottdorf@gmail.com)
+
+package tracer
+
+import (
+	"bytes"
+	"fmt"
+	"math"
+	"strings"
+	"text/tabwriter"
+	"time"
+
+	"github.com/cockroachdb/cockroach/util"
+)
+
+// A Traceable object has a Trace identifier attached to it.
+type Traceable interface {
+	// TraceID is the unique ID for the tracee.
+	TraceID() string
+	// TraceName is a short and "sufficiently unique" human-readable
+	// representation of the tracee.
+	TraceName() string
+}
+
+// A TraceItem is an entry in a Trace.
+type TraceItem struct {
+	depth      int32
+	Origin     string
+	Name       string
+	Timestamp  time.Time
+	Duration   time.Duration
+	Func, File string
+	HLC        time.Time // TODO HLC timestamp
+}
+
+// A Trace is created by a Tracer and records the path of a request within (a
+// connected part of) the system. It contains the ID of the traced object and a
+// slice of trace entries. In typical usage the Epoch() and Event() methods are
+// called at various stages to record the path the associated request takes
+// through the system; when the request goes out of scope, a call to Finalize
+// marks the end of the Trace, at which point it publishes itself to an
+// associated `util.Feed`. A request may create multiple Traces as it passes
+// through different parts of a distributed systems.
+// A Trace is not safe for concurrent access.
```

Comment:
How are you guaranteeing that concurrent access does not occur? For example, are traces local to a goroutine?